py_proxy/views.py

```diff
@@ -50,8 +52,9 @@ def pdf(request):
 @view.view_config(route_name="content_type")
 def content_type(request):
     """Routes the request according to the Content-Type header."""
-    url = request.matchdict["url"]
-    with requests.get(url, stream=True) as rsp:
+    # Drop / from url.
+    url = request.path_qs[1:]
```

Comment:
I'm not particularly proud of this index based approach but I thought it was simpler than piecing the url back together.

Previously we were getting the path from matchdict. However, in order to be complete we would have to generate a url with that and the query params which would most likely involve using urllib3 and it seems kinda silly and complex to me to break apart the url into pieces and then put it back together into the original url. 

Another option

Reply:
Can you also add a unit test for this one, that the query string is preserved on `url`? There should be tests for both when the request does and does not have a query string.

I think this is fine, and not as bad as the similar `/pdf/` truncating above. You could use the same `request.query_string`-based approach as I suggested above though, and it could be moved into a helper method:

```diff
diff --git a/py_proxy/views.py b/py_proxy/views.py
index 845527c..3851c2f 100644
--- a/py_proxy/views.py
+++ b/py_proxy/views.py
@@ -52,8 +52,7 @@ def pdf(request):
 @view.view_config(route_name="content_type")
 def content_type(request):
     """Routes the request according to the Content-Type header."""
-    # Drop / from url.
-    url = request.path_qs[1:]
+    url = append_query_string(request.matchdict["url"], request)
     with requests.get(url, stream=True, allow_redirects=True) as rsp:
         if rsp.headers.get("Content-Type") in ("application/x-pdf", "application/pdf",):
             return exc.HTTPFound(request.route_url("pdf", pdf_url=url))
@@ -61,6 +60,16 @@ def content_type(request):
     return exc.HTTPFound(f"{via_url}/{url}")
 
 
+def append_query_string(url, request):
+    """Return ``url`` with the query string from ``request`` appended.
+
+    If ``request`` has no query string then just return ``url`` unmodified.
+    """
+    if request.query_string:
+        return url + "?" + request.query_string
+    return url
+
+
 def includeme(config):
     """Pyramid config."""
     config.add_route("index", "/")
```

Reply:
I'm currently trying to decide how to actually test this since right now we are mocking everything so any test I wrote would in no way guarantee that the mocked argument return actually contained the query params when not mocked. I can write a test that tests both of them but since we are mocking the value of pdf url if someone unknowingly switched back to matchdict and just changed the mock to return the pdf url, the tests would still pass even though matchdict doesn't include the query params.

Reply:
It doesn't feel like a very strong tests to me. Wondering if there's a better way to test this?

Reply:
The general guideline for our Python unit tests is that:

1. In general when one of our modules uses _another one of our modules_ then you _should_ mock the other module. We test each of our modules in isolation from our own other modules, using mocking when necessary to achieve this isolation.

2. But in general when one of our modules uses a third-party or standard library module you should _not_ mock that third-party module. So for example our model tests run against a real SQLAlchemy and Postgres DB, we don't mock SQLAlchemy. Or in your tests here, you're using httpretty to fake network responses rather than mocking `requests`. Similarly, _tests should not mock the Pyramid request_ or parts of the request. They should either use the `DummyRequest` class that Pyramid provides for testing or use a real Pyramid request

Here's an example of how to test this. I think this is a full-fidelity test that really tests the view code using a real `Request` object and the real `route_url()`. Some notes on this example:

* You wouldn't really write it as a separate `TestContentTypeWithAndWithoutQueryString` class like the example does. You'd just add it into `TestContentTypeRoute`

* You wouldn't really import `Request` and `testing` in the test method (you'd move those imports to the top of the file)

* We don't want real network requests to accidentally be sent if a test forgets to activate httpretty or forgets to register a URI with httpretty. We also don't want httpretty boilerplate (activating it, registering URIs) to be in our tests, we want that tidied away into fixtures. For an example of how to do this see [the `httpretty_` fixture in LMS's `conftest.py`](https://github.com/hypothesis/lms/blob/a860bc44807e6135f5b498c36c74a3b1ff2a4445/tests/unit/conftest.py#L158-L189) and then see [an example test fixture that registers a response](https://github.com/hypothesis/lms/blob/f7649b8da9703488d48c9acbf771ab564c3b1b8f/tests/unit/lms/services/exceptions_test.py#L189-L199). For now, in my example below, the httpretty stuff is just done in the test itself

* Normally tests would use [Pyramid's `DummyRequest`](https://docs.pylonsproject.org/projects/pyramid/en/latest/api/testing.html#pyramid.testing.DummyRequest), which is a dummy request object (not a mock) but [`DummyRequest` doesn't support `path_qs`](https://github.com/Pylons/pyramid/blob/497e667e41b83193d7c5a4f74e9419320d755c46/src/pyramid/testing.py#L361). Or rather, `path_qs` is always an empty string with `DummyRequest`'s.

  This means that, if the code reads `request.path_qs`, the tests for that code would need to manually set `pyramid_request.path_qs = "<PATH_WITH_QUERY_STRING>"`. 

  This isn't a big problem but it does couple the tests to internal implementation details of the code.

  For this reason I've used a real Pyramid `Request` object (constructed by calling [`Request.blank()`](https://docs.pylonsproject.org/projects/pyramid/en/latest/api/request.html#pyramid.request.Request.blank)) in my example below. This sets `path_qs` and most other request attributes automatically.

* The code calls `request.route_url()`. Rather than mocking `route_url()` I've used [`with testing.testConfig(request=request) as config:`](https://docs.pylonsproject.org/projects/pyramid/en/latest/api/testing.html#pyramid.testing.testConfig) and then set up the route with `config.add_route("pdf", "/pdf/{pdf_url:.*}")`. This way the code can call the real `route_url()` and get the expected result. This works with both real and `DummyRequest` objects. In general you shouldn't mock request methods like `route_url()`.

  You probably want something global (a fixture in `conftest.py`) to call the real code that does the `config.add_route()`'s, rather than duplicating the `add_route()` call in the test as I've done in the example.

```python
class TestContentTypeWithAndWithoutQueryString:
    @pytest.mark.parametrize(
        "requested_path,expected_location",
        [
            # If the requested URL has no query string then it should just
            # redirect to the requested URL, with no query string (but with the
            # leading "/" truncated).
            ("/https://example.com/foo", "http://localhost/pdf/https://example.com/foo"),

            # If the requested URL has a query string then the query string
            # should be preserved in the URL that it redirects to.
            ("/https://example.com/foo?bar=baz", "http://localhost/pdf/https://example.com/foo?bar=baz"),
        ],
    )
    @httpretty.activate
    def test_redirect_location(self, requested_path, expected_location):
        from pyramid.request import Request
        from pyramid import testing

        self.register_uri()
        with testing.testConfig() as config:
            config.add_route("pdf", "/pdf/{pdf_url:.*}")
            request = Request.blank(requested_path)

            redirect = views.content_type(request)

        assert redirect.location == expected_location

    @staticmethod
    def register_uri():
        httpretty.register_uri(
            httpretty.GET,
            "https://example.com/foo",
            body="{}",
            adding_headers={"Content-Type": "application/pdf"},
        )

```

Note that there's a tradeoff between making the tests more realistic by using a real `Request` and setting up things like `path_qs` and `route_url()` in a real way, vs keeping the test code simpler and easier to read and write.

`request.matchdict` is a good example of this. Even with the real `Request` object used in my code above, `request.matchdict` is still `{}`. There might be a way for the tests to trigger the necessary Pyramid code to set `matchdict` in the real way. But it would be a lot simpler for the tests to just create either a real `Request` or a `DummyRequest` and then just do `pyramid_request.matchdict = {"url": "..."}`. This does couple the tests a little bit to an implementation detail of the code under test. But it's also a lot more straightforward